distribution/lib/Standard/Base/0.0.0-dev/src/Error/Common.enso

```diff
@@ -33,3 +183,379 @@ type Wrapped_Dataflow_Error payload
+                            True -> handler caught_panic
+                            False -> Panic.throw caught_panic
+                    _ -> Panic.throw caught_panic
+
+    ## Executes the provided action and converts a possible panic matching any of
+       the provided types into a dataflow Error.
+
+       If action executes successfully, the result of `Panic.recover` is the result
+       of that action. Otherwise, if it panicked with a type matching one of the
+       expected error types, that panic is returned as a dataflow error. Unexpected
+       panics are passed through as-is. it is the panic that was thrown after
+       conversion to a dataflow error.
+
+       Arguments:
+       - expected_types: The types of expected panics which should be recovered.
+         This can either be a Vector of types or a single type.
+       - action: The code to execute that potentially panics.
+
+       > Example
+         Converting an expected panic to a dataflow error.
+
+             Panic.recover Illegal_Argument_Error (Panic.throw (Illegal_Argument_Error "Oh!" Nothing))
+
+       > Example
+         Converting one of many expected panic types to a dataflow error.
+
+             Panic.recover [Illegal_Argument_Error, Illegal_State_Error] (Panic.throw (Illegal_Argument_Error "Oh!" Nothing))
+    recover : (Vector.Vector Any | Any) -> Any -> Any
+    recover expected_types ~action =
+        types_to_check = case expected_types of
+            Vector.Vector _ -> expected_types
+            _ -> [expected_types]
+        Panic.catch Any action caught_panic->
+            is_matched = types_to_check.exists typ->
+                caught_panic.payload.is_a typ
+            case is_matched of
+                True -> caught_panic.convert_to_dataflow_error
+                False -> Panic.throw caught_panic
+
+## The runtime representation of a syntax error.
+
+   Arguments:
+   - message: A description of the erroneous syntax.
+@Builtin_Type
+type Syntax_Error message
+
+## The runtime representation of a type error.
+
+   Arguments:
+   - expected: The expected type at the error location.
+   - actual: The actual type at the error location.
+   - name: The name of the argument whose type is mismatched.
+@Builtin_Type
+type Type_Error expected actual name
+
+## The runtime representation of a compilation error.
+
+   Arguments:
+   - message: A description of the erroneous state.
+@Builtin_Type
+type Compile_Error message
+
+## The error thrown when a there is no pattern to match on the scrutinee.
+
+   Arguments:
+   - scrutinee: The scrutinee that failed to match.
+@Builtin_Type
+type Inexhaustive_Pattern_Match_Error scrutinee
+
+## The error thrown when the number of arguments provided to an operation
+   does not match the expected number of arguments.
+
+   Arguments:
+     - expected: the expected number of arguments.
+     - actual: the actual number of arguments passed.
+@Builtin_Type
+type Arity_Error expected actual
```

Comment:
This is not in line with how the arity error was constructed in code- see: `params = {"expected_min", "expected_max", "actual"}`.

```suggestion
   Arguments:
     - expected_min: the minimum expected number of arguments.
     - expected_max: the maximum expected number of arguments.
     - actual: the actual number of arguments passed.
@Builtin_Type
type Arity_Error expected_min expected_max actual
```

I tried to update it in my PR but was unrelated and blocked me from merging, probably would conflict with your change anyway so was pointless for me.

I know you wanted to avoid too many modifications but its a simple one and gets the definition correctly aligned (actually with the new approach shouldn't such a misalignment be reported as an error? If not, what would it take to detect such erroneous conditions?)

Reply:
sgtm